### DISCUSSION TITLE
CslaGenFork

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#4372] [Questions]
I downloaded the CslaGenfork and tried to run it on my machine but I am getting errors in this  [file](https://github.com/MarimerLLC/CslaGenFork/blob/main/branches/V4-3-RC/Solutions/CslaGenFork/Util/ConfigTools.cs).
It's referencing 'SharpeAppConfig' and looking for Mru string

Kind Regards

**Comment by mtavares628** [comment]
I don't think you want to run it out of that branch folder, but rather out of the trunk folder: trunk/Solutions/CslaGenFork.  

I created my own fork of CslaGenFork a while ago and have updated the templates to work with CSLA 8.  You can check it out here:

https://github.com/mtavares628/CslaGenFork

Full disclosure, though.  I've only tailored it to my needs, so the Factory Methods on the BOs pass in the ApplicationContext (or DataPortalFactory) instead of creating a Factory Class for each BO. Also, I'm using regular ADO.NET for DAL generation.

**Reply by Chicagoan2016** [reply]
That's great! I use stored procs with ADO.NET as well. I am going to take a look at your fork.
Kind regards

**Reply by Chicagoan2016** [reply]
I downloaded your code and now I am getting error in 
```cs
namespace CslaGenerator.Util
{
    /// <summary>
    /// Methods for manipulating the key/value pair settings in the app.config file 
    /// (rather the executable.exe.config file).
    /// </summary>
    public static class ConfigTools
```


![ErrorinGenFork](https://github.com/user-attachments/assets/3cbf4a2c-03fd-4dab-9e38-5e277c0862c4)

**Reply by rockfordlhotka** [reply]
> I created my own fork of CslaGenFork a while ago and have updated the templates to work with CSLA 8. You can check it out here:

I must be mis-remembering. Didn't you do a PR to merge that back into the main repo?

**Reply by mtavares628** [reply]
@Chicagoan2016 - I didn't write or change that area of code.  However, if you run it without debugging, the app will run, and then when you set up the templates directory (under the Tools menu), it will actually create the file that is missing and that error should go away when debugging.

@rockfordlhotka - I had forgotten that I had merged my changes into your repo. I did make  some minor tweaks to mine since the last time I submitted that PR. Would you like me to submit another one to bring up to date?

**Reply by rockfordlhotka** [reply]
> @rockfordlhotka - I had forgotten that I had merged my changes into your repo. I did make some minor tweaks to mine since the last time I submitted that PR. Would you like me to submit another one to bring up to date?

Yes, if you don't mind. I think it would be ideal to keep the main repo as up to date as possible.

**Comment by Chicagoan2016** [comment]
@mtavares628  Do we have a short 'getting started' somewhere?
Regards

**Reply by mtavares628** [reply]
There used to be a wiki for it years ago, but it looks like that site doesn't exist anymore.   The initial setup involves the following:

1. Download the Codesmith 2.6 Engine here: https://www.codesmithtools.com/freeware and unzip it somewhere on your machine.
2. Go to Tools --> Codesmith Extension, and point it to where you dropped the Codesmith files.
3.  Go to Tools --> Locate Template Directory and point it to the templates folder under the trunk folder of the solution.
4. Create a new Project File under File New.
5. Give it a project name and Output directory (displayed on the left hand side)
6. Connect your Database from Database --> Connect
7. You will probably want to tweak a bunch of settings under the Project Properties Tab (There are bunch of sub tabs under that). Of Note, under the Generation --> Target Tab, the Target Framework should be CSLA 6.0 DAL. (This will generate CSLA v6+ code with the ADO.NET DAL).
8. Under Creation --> Database defaults the Persistence Type should be Injected SQL Connection.
9. There are other settings as well, but those should get you started.
10. Under the Schema tab you should see your DB tables and Stored Procedures.  You can create a BO from either.  For tables, you would click on the table and see the list of columns in the right pane.  For Stored Procedures, you should need to expand the node for the SP, then right-click and select reload.  This will then add Sub-Nodes for each result set produced by the SP.  Select the Result Set and the list of fields will be displayed in the right pane.
11. Select all of the fields that you want to add to your BO in the Columns Pane, then right-click and you will see a menu that allows you to choose from creating different business objects like editable root, read only list, etc.  Select the one you want and it will prompt you for the class name(s). 
12. The created objects then appear in the left Csla Objects Pane.  When you click on the object, it will update the CslaObject Info pane on the far right. There you can adjust the properties as necessary.  Items of note are Value properties and the Criteria Objects Collection.
13. Make sure that the Value Properties have a field with a primary key (it will set this automatically when creating from a table, but not from a stored procedure).
14. It will create a Criteria Member when you create the BO, but you will need to adjust it.  I would recommend setting the Criteria Class Mode to Multiple Primitives so that it uses primitives for the dataportal methods.  You will also want to fill in the Criteria Properties so set up those parameters.  Then under the Generation options you will need to fill in the Store Procedure names and other settings for the different options.

This should get you started, but there are a lot more settings that you can play around with.

**Reply by Chicagoan2016** [reply]
Thank you Sir. I am going to try it.
Regards
